### ISSUE TITLE
Allow buffers in the spill framework to be freed instead of spilled

### ISSUE THREAD

**Issue by abellina** [NVIDIA/spark-rapids#7709]

**Comment by abellina** [comment]
I have been looking at this a bit and I was thinking we would like to cache `DeviceMemoryBuffer`, but one of the applications that we have (talking to @ttnghia, specifically around caching `GpuJSONPaths`) would need really to cache an `AutoCloseable`, and instead of implementing methods that return a device buffer specifically, we could return some generic type `T`.

Other applications of this are: caching broadcasts, caching build side of joins, and caching null/scalar columns in expand and generally in projections.

The application @ttnghia had in mind won't be ready for a while, so I may look at the null/scalar columns or the broadcast.

**Comment by jlowe** [comment]
Agree, we should try to move away from a specific instance to handle more nebulous cases that might not fit in a DeviceMemoryBuffer mold where we have to know the address and the size.  We need an interface to abstract the concepts the spill framework requires, which is probably a relatively short list, e.g.:
- size (possibly approximate)
- how to "spill" to free device memory
- how to "unspill" if the data is requested and we have previously "spilled"

Some implementations will implement spill and unspill via copying to/from another store.  Others will simply free device memory to "spill" because the data can be recomputed (i.e.: it's a cache of data on device memory).
